File: server/schemas/character.py
from typing import Optional
from pydantic import BaseModel, Field


class CharacterUpdateSchema(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=50)
    title: Optional[str] = Field(None, max_length=50)
    theme: Optional[str] = Field(None, max_length=50)
    avatar: Optional[str] = Field(None, max_length=255)


class HistoryEntrySchema(BaseModel):
    amount: int = Field(..., ge=0, le=10_000_000)
    type: str = Field(..., max_length=50)
    description: str = Field(..., max_length=255)


class ProgressionSyncSchema(BaseModel):
    total_exp: int = Field(..., ge=0, le=100_000_000)
    level: int = Field(..., ge=1, le=1000)
    power: int = Field(..., ge=0, le=100_000_000)
    rank: str = Field(..., min_length=1, max_length=10)
    history_entry: Optional[HistoryEntrySchema] = None

File: client/src/features/character/types/character.ts
export interface CharacterStats {
  id: string;
  characterId: string;
  strength: number;
  knowledge: number;
  discipline: number;
  focus: number;
  endurance: number;
  recovery: number;
  consistency: number;
}

export interface ProgressHistory {
  id: string;
  characterId: string;
  type: string;
  amount: number;
  description: string;
  createdAt: string | Date;
}

export interface ClassSpecialization {
  id: string;
  name: string;
  baseClass: string;
  tier: number;
  requiredLevel: number;
  description?: string | null;
  icon?: string | null;
}

export interface Character {
  id: string;
  userId: string;
  name: string;
  avatar?: string | null;
  theme?: string | null;
  title?: string | null;
  level: number;
  exp: number;
  power: number;
  rank: string;
  gold: number;
  gems?: number;
  towerTokens?: number;
  availableSP: number;
  streakFreezes?: number;
  activeTitleId?: string | null;
  specializationId?: string | null;
  specialization?: ClassSpecialization | null;
  createdAt: string | Date;
  stats?: CharacterStats | null;
  history?: ProgressHistory[];
  activeBuffs?: {
    id: string;
    buffType: string;
    multiplier: number;
    expiresAt: string;
    chargesRemaining?: number;
  }[];
}
